#### Get Data from the following link: http://files.grouplens.org/datasets
/movielens/ml-20m.zip
#### We will be using the following files for this exercise:
ratings.csv : userId,movieId,rating, timestamp
tags.csv : userId,movieId, tag, timestamp
movies.csv : movieId, title, genres

In [2]:
import pandas as pd 

1. Read the dataset using pandas.

In [3]:
# Soln
ratings = pd.read_csv("/home/chandan/Desktop/ASSIGNMENT/session6/ml-20m/ratings.csv") 
tags = pd.read_csv("/home/chandan/Desktop/ASSIGNMENT/session6/ml-20m/tags.csv") 
movies = pd.read_csv("/home/chandan/Desktop/ASSIGNMENT/session6/ml-20m/movies.csv") 

2. Extract the first row from tags and print its type.

In [51]:
tags.head(1)
# tags.head(1).info()
# 

userId  movieId          tag   timestamp
0      18     4141  Mark Waters  1240597180

In [52]:
tags.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 465564 entries, 0 to 465563
Data columns (total 4 columns):
userId       465564 non-null int64
movieId      465564 non-null int64
tag          465548 non-null object
timestamp    465564 non-null int64
dtypes: int64(3), object(1)
memory usage: 14.2+ MB


Output:
    
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 465564 entries, 0 to 465563
Data columns (total 4 columns):
userId       465564 non-null int64
movieId      465564 non-null int64
tag          465548 non-null object
timestamp    465564 non-null int64
dtypes: int64(3), object(1)
memory usage: 14.2+ MB

3. Extract row 0, 11, 2000 from tags DataFrame.

In [53]:
tags.take([0,11,2000])

userId  movieId                tag   timestamp
0         18     4141        Mark Waters  1240597180
11        65     1783      noir thriller  1368149983
2000     910    68554  conspiracy theory  1368043943

OUTPUT:
	userId	movieId	tag	timestamp
0	18	4141	Mark Waters	1240597180
11	65	1783	noir thriller	1368149983
2000	910	68554	conspiracy theory	1368043943

4. Print index, columns of the DataFrame.

In [56]:
tags.index

RangeIndex(start=0, stop=465564, step=1)

In [57]:
tags.index

RangeIndex(start=0, stop=465564, step=1)

5. Calculate descriptive statistics for the 'ratings' column of the
ratings DataFrame. Verify using describe().

In [58]:
rating_df = ratings.rating
print('count',rating_df.count())
print('std',rating_df.std())
print('min',rating_df.min())
print('mean',rating_df.mean())
print('median',rating_df.median())
print('mode',rating_df.mode())
print('max',rating_df.max())

count 20000263
std 1.051988919275684
min 0.5
mean 3.5255285642993797
median 3.5
mode 0    4.0
dtype: float64
max 5.0


In [59]:
rating_df.describe()

count    2.000026e+07
mean     3.525529e+00
std      1.051989e+00
min      5.000000e-01
25%      3.000000e+00
50%      3.500000e+00
75%      4.000000e+00
max      5.000000e+00
Name: rating, dtype: float64

6. Filter out ratings with rating > 5

In [60]:
ratings[ratings.rating > 5]

Empty DataFrame
Columns: [userId, movieId, rating, timestamp]
Index: []

7. Find how many null values, missing values are present. Deal
with them. Print out how many rows have been modified.

In [62]:
# tags.isnull().values.ravel().sum()
tags.isnull().sum()


userId        0
movieId       0
tag          16
timestamp     0
dtype: int64

In [63]:
tags=tags.fillna(" ")

#### 8. Filter out movies from the movies DataFrame that are of type
'Animation'.

In [65]:
# movies['Animation' in movies.genres]
movies["Indexes"]= movies["genres"].str.find('Animation')
movies[movies.Indexes > 0]['title'].head(5)

0                                Toy Story (1995)
12                                   Balto (1995)
360                         Lion King, The (1994)
388    Secret Adventures of Tom Thumb, The (1993)
553                        Pagemaster, The (1994)
Name: title, dtype: object

9. Find the average rating of movies.

In [66]:
ratings['rating'].mean()

3.5255285642993797

# 10. Perform an inner join of movies and tags based on movieId.

In [67]:
merge= pd.merge(movies,tags,how='inner',on='movieId')
merge.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   Indexes  userId                                     tag   timestamp  
0       10    1644                                 Watched  1417736680  
1       10    1741                      computer animation  1183903155  
2       10    1741                 Disney animated feature  1183933307  
3       10    1741                         Pixar animation  1183934770  
4       10    1741  Téa Leoni does not star in this movie  1245093573

# 11. Print out the 5 movies that belong to the Comedy genre and
have rating greater than 4.

In [ ]:

movie_ratings= pd.merge(ratings,movies,how='inner',on='movieId')
movie_ratings.head(10)
# movie_ratings.loc[(movie_ratings["genres"].str.find('Comedy')) & (movie_ratings["rating"] > 4), "title"].head(5)
# query('genres > 50 and C != 900')

1     Toy Story (1995)
4     Toy Story (1995)
7     Toy Story (1995)
9     Toy Story (1995)
14    Toy Story (1995)
Name: title, dtype: object

# 12. Split 'genres' into multiple columns.

In [26]:



genres_split = movies['genres'].str.split("|", expand = False)
genres_split.head(5)
# movies['genres_split'].head(5)

0    [Adventure, Animation, Children, Comedy, Fantasy]
1                       [Adventure, Children, Fantasy]
2                                    [Comedy, Romance]
3                             [Comedy, Drama, Romance]
4                                             [Comedy]
Name: genres, dtype: object

# 13. Extract year from title e.g. (1995).

In [4]:
# movies['a'],movies['b'] = movies['title'].str.split('(', 1).str
# movies['title'].str.split(' ')
new = movies['title'].str.split(" ", expand = True)
new

0            1            2          3        4       5   \
0             Toy        Story       (1995)       None     None    None   
1         Jumanji       (1995)         None       None     None    None   
2        Grumpier          Old          Men     (1995)     None    None   
3         Waiting           to       Exhale     (1995)     None    None   
4          Father           of          the      Bride     Part      II   
5            Heat       (1995)         None       None     None    None   
6         Sabrina       (1995)         None       None     None    None   
7             Tom          and         Huck     (1995)     None    None   
8          Sudden        Death       (1995)       None     None    None   
9       GoldenEye       (1995)         None       None     None    None   
10       American   President,          The     (1995)     None    None   
11       Dracula:         Dead          and     Loving       It  (1995)   
12          Balto       (1995)         None       None     None    None   
13          Nixon       (1995)         None       None     None    None   
14      Cutthroat       Island       (1995)       None     None    None   
15         Casino       (1995)         None       None     None    None   
16          Sense          and  Sensibility     (1995)     None    None   
17           Four        Rooms       (1995)       None     None    None   
18            Ace     Ventura:         When     Nature    Calls  (1995)   
19          Money        Train       (1995)       None     None    None   
20            Get       Shorty       (1995)       None     None    None   
21        Copycat       (1995)         None       None     None    None   
22      Assassins       (1995)         None       None     None    None   
23         Powder       (1995)         None       None     None    None   
24        Leaving          Las        Vegas     (1995)     None    None   
25        Othello       (1995)         None       None     None    None   
26            Now          and         Then     (1995)     None    None   
27     Persuasion       (1995)         None       None     None    None   
28           City           of         Lost  Children,      The   (Cité   
29       Shanghai        Triad         (Yao          a      yao     yao   
...           ...          ...          ...        ...      ...     ...   
27248      Werner            -       Volles     Rooäää   (1999)    None   
27249        What            A          Man     (2011)     None    None   
27250           7     Dwarves:          The     Forest       Is     Not   
27251         The          Fat          Spy     (1966)     None    None   
27252         Die  Bademeister            –    Weiber,  saufen,   Leben   
27253        Ants           in          the      Pants        2  (2002)   
27254      Manta,        Manta       (1991)       None     None    None   
27255       Oscar          and          the       Lady       in    Pink   
27256         Por           un       puñado         de    besos  (2014)   
27257     Vietnam           in           HD     (2011)     None    None   
27258        WWII           IN           HD     (2009)     None    None   
27259     Phoenix       (2014)         None       None     None    None   
27260   Parallels       (2015)         None       None     None    None   
27261      Closed      Curtain       (2013)       None     None    None   
27262   Gentlemen       (2014)         None       None     None    None   
27263           A       Second       Chance     (2014)     None    None   
27264        Dead      Rising:   Watchtower     (2015)     None    None   
27265     Standby       (2014)         None       None     None    None   
27266        What          Men         Talk      About   (2010)    None   
27267       Three      Quarter         Moon     (2011)     None    None   
27268        Ants           in          the      Pants   (2000)    None   
27269      Werner            -      Geko